src/rt/sched/behaviourcore.h

```diff
@@ -57,74 +57,333 @@ namespace verona::rt
 
   struct BehaviourCore;
 
+  inline Logging::SysLog& operator<<(Logging::SysLog&, BehaviourCore&);
+
   struct Slot
   {
-    Cown* cown;
+  private:
+    /**
+     * Cown required by this behaviour
+     *
+     * Bit 0 - Possible values:
+     *          0 - Wait (2PL in acquire phase going on)
+     *          1 - Ready (2PL in acquire phase done)
+     *
+     * Bit 1 - Possible values:
+     *          0 - Current slot Writer
+     *          1 - Current slot Reader
+     *
+     * Remaining bits - Cown pointer
+     *
+     * Assumption - Cowns are allocated at 4 byte boundary. Last 2 bits are
+     * zero.
+     */
+    std::atomic<uintptr_t> _cown;
+
+    static constexpr uintptr_t WAIT_FLAG = 0x0;
+    static constexpr uintptr_t READY_FLAG = 0x1;
+    static constexpr uintptr_t STATUS_BITS = 0x1;
+    static constexpr uintptr_t STATUS_SHIFT = 0x0;
+
+    static constexpr uintptr_t CURR_SLOT_WRITER_FLAG = 0x0;
+    static constexpr uintptr_t CURR_SLOT_READER_FLAG = 0x1;
+    static constexpr uintptr_t CURR_SLOT_TYPE_SHIFT =
+      STATUS_BITS + STATUS_SHIFT;
+    static constexpr uintptr_t CURR_SLOT_TYPE_BITS = 0x1;
+
+    static constexpr uintptr_t COWN_SHIFT =
+      CURR_SLOT_TYPE_SHIFT + CURR_SLOT_TYPE_BITS;
+
     /**
-     * Possible values before scheduling to communicate memory management
-     * options:
-     *   0 - Borrow
-     *   1 - Move
-     *
-     * Possible vales after scheduling:
-     *   0 - Wait
-     *   1 - Ready
-     *   Behaviour* - Next write
-     *
-     * TODO Read-only When we extend to read-only we will need the following
-     * additional state
-     *   Slot* - Next Read
-     *   2 - Ready and Read available
+     * Next slot in the MCS Queue
+     *
+     * Bit 0 - Possible values:
+     *          1 - Current slot active
+     *          0 - Current slot blocked
+     *
+     * Bit 1 - Possible values:
+     *          0 - Next slot Writer
+     *          1 - Next slot Reader
+     *
+     * Remaining bits  - Next slot pointer
+     *
+     * Before scheduling: Bit 0 => Whether move or not
+     *
+     * Assumption - Slots are allocated at 4 byte boundary. Last 2 bits are
+     * zero.
      */
     std::atomic<uintptr_t> status;
 
-    Slot(Cown* cown) : cown(cown), status(0) {}
+    static constexpr uintptr_t SLOT_BLOCKED_FLAG = 0x0;
+    static constexpr uintptr_t SLOT_ACTIVE_FLAG = 0x1;
+    static constexpr uintptr_t SLOT_ACTIVE_BITS = 0x1;
+    static constexpr uintptr_t SLOT_ACTIVE_SHIFT = 0x0;
+
+    static constexpr uintptr_t NEXT_SLOT_WRITER_FLAG = 0x0;
+    static constexpr uintptr_t NEXT_SLOT_READER_FLAG = 0x1;
+    static constexpr uintptr_t NEXT_SLOT_TYPE_SHIFT =
+      SLOT_ACTIVE_BITS + SLOT_ACTIVE_SHIFT;
+    static constexpr uintptr_t NEXT_SLOT_TYPE_BITS = 0x1;
+
+    static constexpr uintptr_t NEXT_POINTER_SHIFT =
+      NEXT_SLOT_TYPE_BITS + NEXT_SLOT_TYPE_SHIFT;
+
```

Comment:
This feels overly complicated, and seems more likely to introduce bugs from not using the correct constant. Could you just use:

```suggestion
    static constexpr uintptr_t STATUS_SLOT_ACTIVE_FLAG = 0x1;
    static constexpr uintptr_t STATUS_NEXT_SLOT_READER_FLAG = 0x2;
    static constexpr uintptr_t STATUS_NEXT_SLOT_MASK = ~(STATUS_SLOT_ACTIVE_FLAG | STATUS_NEXT_SLOT_READER_FLAG);
```

```C++
    /**
     * Returns true if the next slot wants to acquire in read mode
     */
    bool is_next_slot_read_only()
    {
      assert()
      return ((status.load(std::memory_order_acquire) & STATUS_NEXT_SLOT_READER_FLAG) != 0);
    }

    /**
     * Returns true if the next slot wants to acquire in write mode
     */
    bool is_next_slot_writer()
    {
      return !is_next_slot_read_only();
    }
```

```C++
    /**
     * Return the next slot
     */
    Slot* next_slot()
    {
      return (Slot*)(status.load(std::memory_order_acquire) & STATUS_NEXT_SLOT_MASK);
    }
```